### ISSUE TITLE
RuntimeError: Expected tensor for argument #1 'input' to have the same type as tensor for argument #3 'offset';

### ISSUE THREAD

**Issue by vantienpham** [inspiros/tvdcn#4]
Hi, thanks for this excellent work.
I'm embedding the packed classes of this repo into the [InternImage](https://github.com/OpenGVLab/InternImage/blob/631a5159e2c4e4bda16c732e64fa9584b38859ea/classification/ops_dcnv3/modules/dcnv3.py#L336) architecture.
My plan is to replace the `dcnv3` operator with the `PackedDeformConv2d`/`PackedDeformConvTranspose2d`.
 The input and output shape in [the forward function of dcnv3](https://github.com/OpenGVLab/InternImage/blob/631a5159e2c4e4bda16c732e64fa9584b38859ea/classification/ops_dcnv3/modules/dcnv3.py#L318) is identical (from my verification), but I want to adopt their [projections](https://github.com/OpenGVLab/InternImage/blob/631a5159e2c4e4bda16c732e64fa9584b38859ea/classification/ops_dcnv3/modules/dcnv3.py#L297) so I just replace their conv/offset/mask with this one:
```python
self.conv = PackedDeformConv2d(
            in_channels=channels,
            out_channels=channels,
            kernel_size=kernel_size,
            stride=stride,
            padding=pad,
            dilation=dilation,
            groups=channels,
            modulated=modulated,
            mask_activation=mask_activation,
            dtype=torch.float32,
            device='cuda'
        )
```
Since their `forward` function processes the input in the form of `(N,H,W,C)` so a permutation is performed:
```python
def forward(self, input):
        """
        :param query                       (N, H, W, C)
        :return output                     (N, H, W, C)
        """
        N, H, W, _ = input.shape
        N, H, W, C = input.shape


        x = self.input_proj(input)
        x_proj = x
        dtype = x.dtype

        x1 = input.permute(0, 3, 1, 2)
        # x1 = self.dw_conv(x1)
        # offset = self.offset(x1)
        # mask = self.mask(x1).reshape(N, H, W, self.group, -1)
        # mask = F.softmax(mask, -1)
        # mask = mask.reshape(N, H, W, -1).type(dtype)
        
        # x = DCNv3Function.apply(
        #     x, offset, mask,
        #     self.kernel_size, self.kernel_size,
        #     self.stride, self.stride,
        #     self.pad, self.pad,
        #     self.dilation, self.dilation,
        #     self.group, self.group_channels,
        #     self.offset_scale,
        #     256,
        #     self.remove_center)
        # print(input.dtype)
        # print('input.get_device() ', input.get_device())
        # print(x1.dtype)
        # x2 = input.permute(0, 3, 1, 2).float()
        # x2 = torch.randn(N, C, H, W).to('cuda')
        # print(x2.shape)
        # print(x2.dtype)
        x1 = self.conv(x1)
        
        if self.center_feature_scale:
            center_feature_scale = self.center_feature_scale_module(
                x1, self.center_feature_scale_proj_weight, self.center_feature_scale_proj_bias)
            # N, H, W, groups -> N, H, W, groups, 1 -> N, H, W, groups, _d_per_group -> N, H, W, channels
            center_feature_scale = center_feature_scale[..., None].repeat(
                1, 1, 1, 1, self.channels // self.group).flatten(-2)
            x = x * (1 - center_feature_scale) + x_proj * center_feature_scale
        x = self.output_proj(x)
        return x
``` 
However, when running the code, it yields this error:
```bash
[2023-07-07 11:41:35 internimage_t_1k_224](main.py 307): INFO Start training
torch.Size([32, 56, 56, 64])
TDCN(
  (conv): PackedDeformConv2d(
    64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=64
    (offset_generator): Conv2d(64, 18, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (mask_generator): Conv2d(64, 9, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (mask_activation): MaskSigmoid()
  )
  (input_proj): Linear(in_features=64, out_features=64, bias=True)
  (output_proj): Linear(in_features=64, out_features=64, bias=True)
)
Traceback (most recent call last):
  File "/home/tien/InternImage/classification/main.py", line 661, in <module>
    main(config)
  File "/home/tien/InternImage/classification/main.py", line 312, in main
    train_one_epoch(config,
  File "/home/tien/InternImage/classification/main.py", line 416, in train_one_epoch
    outputs = model(samples)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1156, in forward
    output = self._run_ddp_forward(*inputs, **kwargs)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1110, in _run_ddp_forward
    return module_to_run(*inputs[0], **kwargs[0])  # type: ignore[index]
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/InternImage/classification/models/intern_image.py", line 783, in forward
    x = self.forward_features(x)
  File "/home/tien/InternImage/classification/models/intern_image.py", line 742, in forward_features
    x = level(x)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/InternImage/classification/models/intern_image.py", line 529, in forward
    x = blk(x)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/InternImage/classification/models/intern_image.py", line 447, in forward
    x = _inner_forward(x)
  File "/home/tien/InternImage/classification/models/intern_image.py", line 433, in _inner_forward
    x = x + self.drop_path(self.dcn(self.norm1(x)))
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/InternImage/classification/ops_dcnv3/modules/dcnv3.py", line 525, in forward
    x1 = self.conv(x1)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/tvdcn/ops/deform_conv.py", line 631, in forward
    return self._conv_forward(input, self.weight, offset, mask, self.bias)
  File "/home/tien/anaconda3/lib/python3.9/site-packages/tvdcn/ops/deform_conv.py", line 345, in _conv_forward
    return deform_conv2d(input, weight, offset, mask, bias,
  File "/home/tien/anaconda3/lib/python3.9/site-packages/tvdcn/ops/deform_conv.py", line 141, in deform_conv2d
    return torch.ops.tvdcn.deform_conv2d(
  File "/home/tien/anaconda3/lib/python3.9/site-packages/torch/_ops.py", line 502, in __call__
    return self._op(*args, **kwargs or {})
RuntimeError: Expected tensor for argument #1 'input' to have the same type as tensor for argument #3 'offset'; but type torch.cuda.FloatTensor does not equal torch.cuda.HalfTensor (while checking arguments for deform_conv2d_forward)
```
Why the `dtype` of `offset` in this case is HalfTensor?

I tried to use the `PackedDeformConv2d` in a common bash with a similar input shape as the code above, it worked smoothly:
```bash
>>> x.shape
torch.Size([32, 64, 56, 56])
>>> Tconv
PackedDeformConvTranspose2d(
  64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=64
  (offset_generator): Conv2d(64, 18, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (mask_generator): Conv2d(64, 9, kernel_size=(1, 1), stride=(1, 1), bias=False)
  (mask_activation): MaskSigmoid()
)
>>> Tconv(x).shape
torch.Size([32, 64, 56, 56])
```
Sorry, I'm unable to provide a MWE at the moment. Any recommendation is appreciated!
Thanks in advance.

**Comment by inspiros** [comment]
Somehow the generated `offset` tensor is of type `torch.half`. I suggest that you should cast the model's parameters and buffers to `torch.float` before doing anything:
```python
model = model.float()
```
Then ensure the input is cast with `input = input.float()` before passing into the model. Let me know if the problem persists.

And why are you here, do you want me to add you as Collaborator 🧐?

**Comment by vantienpham** [comment]
Thank you for your suggestion.
I've attempted to cast both the model and input both outside and inside the forward function, but the problem persists.
If you believe that this issue is unrelated to the repository, please feel free to close it.

**Comment by vantienpham** [comment]
Closing as it doesn't relate to tvdcn.
I'm not sure about the root of the problem but in my case with InterImage's code, it relates to the [mixed precision setting](https://github.com/OpenGVLab/InternImage/blob/7142c933b5dc64f33a88539e12108bdbcce11b5e/classification/main.py#L106).
For future reference: I've tried to covert to the related module/tensors with `.float()` or `.to(dtype=torch.float)` as this [suggestion](https://discuss.pytorch.org/t/how-to-cast-a-tensor-to-another-type/2713/15) but when the amp is on, these instructions does not change the type of the objects.
